more idiomatic Rust parser

remi-dupre opened the pull request:
I changed many things on how the generated Rust parsers behave, maybe we don't want all of these changes.


### Derive traits

I added some auto trait implementations with `#[derive(Clone, Debug, Eq, Hash, PartialEq)]`, it isn't really important but it makes us look fancy and sometime people might forget that they are not here.


### Use a single buffer

Maybe I overengineered here but I think that one of the nice designs of Rust is that it makes you think hard about when you are making memory allocation (typically when using `::new()`, `::clone()`).

There is now one memory allocation for input buffer (plus a few realloc) and exactly one for each list in the input, isn't that cool ? Note that it will probably not change performances in our use case. :(


### In-lining list parsing

I think this isn't too heavy for the main function, it also makes it easier to change the integer type for the candidate. This last point is quite important because `i32` is often not very convenient to work with, for example when it has to be compared with `usize`.


### Use FromStr for one-line structs

It makes us stylish to use `string.parse()` to read a struct. I also think it is really way better to work with traits from standard library when there is one corresponding to our use case.


### Remove all `unwrap()`

I think that `unwrap()` should never be used in a clean codebase, except in places where we are sure that no error can be raised. Plus I don't like how competitive programming people always write extremely short and extremely ugly code, so I like the idea of pushing in the total opposite direction :D


### Use of `collect()` to build vectors

It is just more idiomatic, I think it generates exactly the same binary here.


**NOTE**: I wasn't able to run tests, I might need some help with that :(
**NOTE**: I really didn't enjoy this Python formater :'(

Shaac commented:
Hi, thanks for the PR.

I’ll make some general comments here about the end result, and I’ll review the actual python code after.

> ### Derive traits

This looks nice. If these are common traits to derive, it is a good thing that we derive them. Let’s not forget one goal of Iorgen is to give some good examples for the candidates to follow when they code other stuff.

> ### Use a single buffer

I don’t know rust enough to have an opinion on this one. One small issue though, the name buffer will conflict with any variable the user may want to call `buffer`. We have to may sure there is no conflicts (there are several mechanisms in Iorgen for that).

> ### In-lining list parsing

I probably made functions of those because I did not manage to compile with them in the main function. But if you success, the main function is a good place for it. The rust generator is the only one in Iorgen with that many additional functions.

> ### Use FromStr for one-line structs

Ok if you think it is better. Do we have to use `use std::str::FromStr`, or can we keep in inside the namespace? That way it would not conflict with a possible struct from the user named `FromStr`.

> ### Remove all `unwrap()`

So basically you replaced the called to `unwrap()` with calls to `expect`? Or did I miss something. In that case this seems a good idea indeed.

> ### Use of `collect()` to build vectors

Ok

> **NOTE**: I wasn't able to run tests, I might need some help with that :(

What part of the process to run the tests was not clear? I’ll improve that part of the doc.
To run them just go to the `test` folder and run `./test.py -l rs`. And they are failing (the generated code does not compile)

> **NOTE**: I really didn't enjoy this Python formater :'(

Do you have a better suggestion?


On other thing while you are at it. When I implemented the parser I was concerned: when a `Vec` is read, and then given to the user function to be completed, don’t we make an unnecessary copy? Is there a way to take the vec by reference? And if yes, for what types should we do this? strings as well?

remi-dupre commented:
Hi @Shaac!

It has been quite a while but I finally spent time on this PR, I think I fixed most of the issues \o/

> I probably made functions of those because I did not manage to compile with them in the main function. But if you success, the main function is a good place for it. The rust generator is the only one in Iorgen with that many additional functions.

We could actually safely inline every function calls into the main function but it would probably be quite harder to read:

```rust
fn f() -> String {
    // Some computation
}

fn main() {
    let x1 = f();
    let x2 = {
        // Some computation
        // This is safely executed in a separate scope
    };
    assert_eq!(x1, x2);
}
```

> Ok if you think it is better. Do we have to use `use std::str::FromStr`, or can we keep in inside the namespace? That way it would not conflict with a possible struct from the user named `FromStr`.

I changed this to avoid all module imports in general, this may also help the candidate by starting the file directly with relevant code (struct definitions and the function that has to be filled).

> So basically you replaced the called to `unwrap()` with calls to `expect`? Or did I miss something. In that case this seems a good idea indeed.

That's actually roughly what I did, I think the guideline is to avoid using `unwrap()` unless you have a very good reason to think that it will always succeed (eg: you just checked that an `Option` contains a value, but you only unwrap it later for some reason). `expect` is the next quick solution that encourages to explain why the piece of code may fail.

I tweaked this a bit to try to find a balance between compacity and what I would subjectively consider a "clean" error handling.


> > **NOTE**: I wasn't able to run tests, I might need some help with that :(
> 
> What part of the process to run the tests was not clear? I’ll improve that part of the doc.
> To run them just go to the `test` folder and run `./test.py -l rs`. And they are failing (the generated code does not compile)

I don't remember what I struggled with back then, I didn't have any issue when I went back at it, and it was actually really pleasant to work with these tests.

> > **NOTE**: I really didn't enjoy this Python formater :'(
> 
> Do you have a better suggestion?

I'm actually used to [black](https://github.com/psf/black) but using format string appeared to remove all pain I had: I recall my issue was mostly with how it indented long string concatenations and function calls with a large number of arguments. I didn't feel any frustration after this change :)

> On other thing while you are at it. When I implemented the parser I was concerned: when a `Vec` is read, and then given to the user function to be completed, don’t we make an unnecessary copy? Is there a way to take the vec by reference? And if yes, for what types should we do this? strings as well?

Nope, the `Vec` is not copied because Rust uses a move semantic. This makes the function call similar to what I think can be done in  C++ with `std::move` or RVO.

It would be possible to pass a reference, but it would be less flexible than what we currently do :)

Shaac commented on iorgen/parser_rust.py:
```diff
@@ -17,15 +18,19 @@
 
+# KEYWORDS extended with the list of identifiers used by the generated code
+EXTENDED_KEYWORDS = KEYWORDS + [
+    "buffer", "read_line", "std", "Box", "String", "Vec"
```
Shouldn’t you add `Result` to the list? I’ve seen it used a lot and I fear for collision with a user struct.

remi-dupre replied:
Nice catch!

Shaac commented on iorgen/parser_rust.py:
```diff
@@ -17,15 +18,19 @@
+]
+
+# Conventional indentation token
 INDENTATION = "    "
```
You did not use this in most of your code. It’s not a deal breaker because we will probably always use 4 spaces for rust, but you should at least add in your comment that it will not be used everywhere in the code, and that changing the content of `INDENTATION` will not work as expected.

remi-dupre replied:
Oh! That's what I forgot to ask about in my last comment!

Using format strings makes it easy to generate pieces of codes with a quite nice indentation, but I chose to not consistently use the `INDENTATION` constant. If you think this is something we should keep, this can probably be dealt with a fair style such as:

```python
__ = INDENTATION

read_line = [
    u"fn read_line(mut buffer: &mut String) -> &str {",
    f"{__}buffer.clear();",
    f"{__}std::io::stdin()",
    f"{__}{__}.read_line(&mut buffer)",
    f'{__}{__}.expect("impossible to read a new line");',
    f"{__}buffer.trim_end()",
    u"}",
]
```

Shaac replied:
That’s a good idea, but I’m not sure `__` is a variable name that should be used. Maybe something like `tab`

Shaac reviewed (changes requested):
Sorry for the delay in reviewing.
Overall it looks good, I just had a few remarks.

More importantly the CI fails, it should be simple to fix:
```
iorgen/parser_rust.py:84:4: C0116: Missing function or method docstring (missing-function-docstring)
iorgen/parser_rust.py:87:4: C0116: Missing function or method docstring (missing-function-docstring)
iorgen/parser_rust.py:90:8: R1705: Unnecessary "elif" after "return" (no-else-return)
iorgen/parser_rust.py:122:8: R1705: Unnecessary "else" after "return" (no-else-return)
```

remi-dupre commented on test/samples/simple/simple.rs:
```diff
@@ -1,22 +1,27 @@
-    let other_number: i32 = read_line().parse().unwrap();
+    let mut buffer = String::new();
+
+    let n = read_line(&mut buffer)
```
Even though the norm is usually to avoid type annotation whenether you can, I kind of agree with you that it can make it harder to read in this case.

What actually makes me think it can be better in this case is that the candidate can change the type of an input variable by just changing it in the function parameter, at exactly one place. This can be very useful as the strong type system of Rust won't let you use an i32 when another integer type is expected (such as `usize` for indexing arrays).

Shaac replied:
Ok, it is not a big deal if the main function is a bit hard to read, so we can keep it like you did.

remi-dupre commented on test/samples/structs/structs.rs:
```diff
@@ -46,61 +48,81 @@ fn structs(struct_: Struct1, n: i32, struct_list: Vec<Struct1>, triangle: Vec<Po
+        .parse()
+        .expect("invalid `n` parameter");
+
+    let struct_list = (0..n)
```
The type is resolved during the function call, which explicitly requires a `Vec<Struct1>`, do you think it can be very confusing for a beginner or is it just a question you ask yourself as a reviewer?

Shaac replied:
No it’s probably not confusing. For someone that does not know much it may look like magic, but that’s not a real problem.

remi-dupre commented on test/samples/structs/structs.rs:
```diff
@@ -46,61 +48,81 @@ fn structs(struct_: Struct1, n: i32, struct_list: Vec<Struct1>, triangle: Vec<Po
-    let name: char = read_line().parse().unwrap();
-    let pos: Position = read_struct_position();
-    Point { name, pos }
+fn read_struct_point(mut buffer: &mut String) -> Result<Point, Box<dyn std::error::Error>> {
```
The parser creates a function when the structure is defined over several lines (when the struct as a simpler representation it implements `FromStr` which can be used through the `str.parse()` method).

Shaac replied:
Ok then.

Shaac commented:
The PR is mostly good. Mostly the CI failing. Do you want me to finish and merge it?

remi-dupre commented:
> The PR is mostly good. Mostly the CI failing. Do you want me to finish and merge it?

Yeah, I was thinking about that just yesterday (while seeing all these exercises being pushed). I'll try to take some time for it :)

remi-dupre commented:
It should be fine now :

  - fixed merge conficts (it was only addressing a generated file in tests)
  - fixed the few remaining lints, they were actually disabled by my own pylint configuration, that's why I missed them :/

Shaac reviewed (approved):
This is ready to be merged for me!

Please squash your commits before merging. Either in a single one, or several ones if there is some logic to it, but currently there are too many.

remi-dupre commented:
I just squashed the whole thing :+1:

Shaac commented:
It is merged, thanks!
